Feature/strict types

arueckauer opened the pull request:
|    Q          |   A
|-------------- | ------
| Documentation | no
| Bugfix        | no
| BC Break      | yes
| New Feature   | no
| RFC           | no
| QA            | yes

### Description

This PR provides syntactical strict typing and is the base of further improvements towards strict types.

- Add strict type declaration
- Add return type declaration
- Add argument type declaration (where applicable)
- Update class imports
- Remove assertions and type casts which are made redundant by type declaration
- Fix type errors
- Fix second parameter type in calls of `var_export`
- ...

arueckauer committed:
Add strict type declaration

Signed-off-by: Andi Rückauer <arueckauer@gmail.com>

arueckauer committed:
Fix usage of var_export(), providing now a boolean instead of integer for parameter 2

Signed-off-by: Andi Rückauer <arueckauer@gmail.com>

arueckauer committed:
Add class imports to shorten FQCN usages

Signed-off-by: Andi Rückauer <arueckauer@gmail.com>

arueckauer committed:
Fix incompatible return types
Update @return in DocBlock
Remove superfluous return statement

Signed-off-by: Andi Rückauer <arueckauer@gmail.com>

arueckauer committed:
Declare return types
Update docBlock where applicable
Remove redundant return statements
Remove type checks where applicable
Remove test cases which varified type casting, which is no longer supported because of declared argument types

Signed-off-by: Andi Rückauer <arueckauer@gmail.com>

arueckauer committed:
Fix namespace to match PSR-4 project structure

Signed-off-by: Andi Rückauer <arueckauer@gmail.com>

arueckauer committed:
__construct cannot return a value

Signed-off-by: Andi Rückauer <arueckauer@gmail.com>

arueckauer committed:
Safely remove null assignment

Signed-off-by: Andi Rückauer <arueckauer@gmail.com>

arueckauer committed:
Explicitly declare previously dynamically initialized property

Signed-off-by: Andi Rückauer <arueckauer@gmail.com>

arueckauer committed:
Use class constant instead of FQCN string

Signed-off-by: Andi Rückauer <arueckauer@gmail.com>

arueckauer committed:
Declare argument types
Remove type casts which where made obsolete by type declaration
Remove test cases which where made obsolete by type declaration
Remove docBlocks which do not add valuable information

Signed-off-by: Andi Rückauer <arueckauer@gmail.com>

arueckauer committed:
Remove space before return type declaraion colon

Signed-off-by: Andi Rückauer <arueckauer@gmail.com>

arueckauer committed:
Remove unnecessary variable assignments

Signed-off-by: Andi Rückauer <arueckauer@gmail.com>

arueckauer committed:
Fix type hints for  and

Signed-off-by: Andi Rückauer <arueckauer@gmail.com>

arueckauer committed:
Define access modifier

Signed-off-by: Andi Rückauer <arueckauer@gmail.com>

arueckauer committed:
Remove assertions which were made obsolete by type declarations

Signed-off-by: Andi Rückauer <arueckauer@gmail.com>

arueckauer committed:
Add strict comparison in usage of in_array() for type safety

Signed-off-by: Andi Rückauer <arueckauer@gmail.com>

arueckauer committed:
Add strict type comparison

Signed-off-by: Andi Rückauer <arueckauer@gmail.com>

arueckauer committed:
Remove unused import

Signed-off-by: Andi Rückauer <arueckauer@gmail.com>

arueckauer committed:
Remove docBlocks which do not add valuable information
Add missing @throws tags
Add/remove import statements to match current usage of classes

Signed-off-by: Andi Rückauer <arueckauer@gmail.com>

arueckauer committed:
Fix argument type error caused by Reflection::getDocComment(), which might return false (instead of assumed string)

Signed-off-by: Andi Rückauer <arueckauer@gmail.com>

arueckauer committed:
Remove redundant @throws tags

Signed-off-by: Andi Rückauer <arueckauer@gmail.com>

arueckauer committed:
Remove no longer used @group tags

Signed-off-by: Andi Rückauer <arueckauer@gmail.com>

samsonasik commented on src/Definition.php:
```diff
@@ -6,69 +6,55 @@
 
-        if (is_numeric($name)) {
-            $name = null;
-        }
```
string can be numeric, eg : 

```php
➜  ~ php -a
Interactive shell

php > var_dump(is_string('1'));
php shell code:1:
bool(true)

php > var_dump(is_numeric('1'));
php shell code:1:
bool(true)
```

arueckauer replied:
ACK. Add check for numeric string and reset if so (see 58c9e97).

arueckauer committed:
Remove unnecessary FQCN

Signed-off-by: Andi Rückauer <arueckauer@gmail.com>

arueckauer committed:
Add check for numeric string and reset  if so
Add test cases for numeric key when adding (a) method(s)

Signed-off-by: Andi Rückauer <arueckauer@gmail.com>

weierophinney commented on src/AbstractServer.php:
```diff
@@ -6,17 +6,17 @@
-     * @var bool Flag; whether or not overwriting existing methods is allowed
+     * @var bool
      */
     protected $overwriteExistingMethods = false;
```
Which PHP version are you targeting for this update? If you were to bump to 7.4, we could do property typehints here, which would make many of these property docblocks completely unnecessary!

arueckauer replied:
Absolutely right, typed property would bring some welcomed improvements. Currently `dev-3.0` is bumped to 7.2. Would a jump to 7.4 be generally be accepted?

weierophinney replied:
Yes! Particularly as there are definitely features that you could use here.

arueckauer replied:
Ok. Then my suggestion would be: 1) merging this PR, 2) bumping to PHP 7.4 and 3) then adding typed properties.

arueckauer replied:
@weierophinney Are you okay with me resolving this point? 😊

weierophinney replied:
Yep - and I'm doing it now. :smile:

weierophinney commented on src/AbstractServer.php:
```diff
@@ -139,15 +127,15 @@ protected function _dispatch(Method\Definition $invokable, array $params)
-        if ('function' == $type) {
+        if ('function' === $type) {
             $function = $callback->getFunction();
             return call_user_func_array($function, $params);
```
This could be refactored to `return $function(...$params);` which is valid in any PHP 7 version.

arueckauer replied:
Same here, I would like to provide this change in another PR.

weierophinney commented on src/AbstractServer.php:
```diff
@@ -139,15 +127,15 @@ protected function _dispatch(Method\Definition $invokable, array $params)
 
-        if ('static' == $type) {
+        if ('static' === $type) {
             return call_user_func_array([$class, $method], $params);
```
I think the `Method\Definition` should simply have a `getCallback()` method. We can then use the construct `return $callback(...$params);` everywhere.

Static callbacks can be returned as either `[$class, $method]` or `{$class}::{$method}`; with PHP 7, a variable in either notation is callable, and can be called in the same way as a function. Instance callbacks can be returned as `[$instance, $method]`, and, again, can be called in the same way.

This way, there's no reason to test for `$type` any more. (This was a remnant from the PHP 5 days, when we had to call callbacks in different ways based on how they were constructed.)

arueckauer replied:
Thanks for the explanation. I quite like the idea.

arueckauer replied:
I gave it some more thought and would like to provide this change in a separate PR, since it is somewhat out of context to strict typing. Would that be alright with you, @weierophinney ?

weierophinney replied:
Absolutely.

arueckauer replied:
ty

weierophinney reviewed (changes requested):
Overall, it's really fun to see how much code can be removed and/or simplified in this component due to advances in the language! The root of this code dates back to January of 2006, believe it or not!

I've noted a few things more that I think we can do around callbacks, since PHP 7 has unified callback invocation now, and because we now have the splat/unpacking operator.

froschdesign commented on src/AbstractServer.php:
```diff
@@ -126,11 +113,12 @@ protected function _buildSignature(Reflection\AbstractFunction $reflection, $cla
     /**
      * Dispatch method
      *
+     * @param Method\Definition $invokable
+     * @param array $params
```
All parameters in DocBlocks can be omitted if there is no more information than in the parameter list of the method itself. The same for the return value.

arueckauer replied:
I left it because of @deprecated and scheduled change for 3.0 to remove underscore prefix.

froschdesign replied:
I think you misunderstood me. I mean the lines with `@param`, these are not needed here.

arueckauer replied:
If I understand you right, you suggest removing @param and @return tags which are handled by type declarations? Would you keep @throws tags?
So far I've either had a complete DocBlock or none. Because incomplete DocBlocks do not validate.
![Missing PHPDoc](https://user-images.githubusercontent.com/1815979/81269590-cfb55000-9049-11ea-9699-7b72dad676ef.jpg)

froschdesign replied:
Search for "Redundant PHPDoc comment" (Inspections) in the preferences of PhpStorm this should solve the problem.

arueckauer replied:
Got it, thank you.

froschdesign replied:
I thank you for your work. 👍

froschdesign commented on src/Cache.php:
```diff
@@ -29,11 +29,12 @@ class Cache
      *
-     * @param  string $filename
-     * @param  \Laminas\Server\Server $server
+     * @param string $filename
+     * @param Server $server
      * @return bool
```
The same here: the exact information is already included in the parameter list and in the return value of the method.

arueckauer replied:
Even if there is a description and also a @throws tag?

froschdesign replied:
No no, omit only the lines with `@param` and `@return` if the information is already present in parameter list and the return value. Descriptions and `@throws` are good and an important information.

(Maybe that wasn't clear enough but I marked only the lines 32 - 34, not the entire DocBlock. I didn't mean to mislead you.)

arueckauer replied:
Probably misunderstood you. Thanks for the explanation.

I went over the codebase and removed all redundant @param and @return tags. I check this point as resolved. If I've overlooked something, please let me know.

arueckauer committed:
Remove is_string() which is made obsolete by argument type declaration

Signed-off-by: Andi Rückauer <arueckauer@gmail.com>

arueckauer committed:
Remove @param and @return tags which are simly a redundant representation of declared types

Signed-off-by: Andi Rückauer <arueckauer@gmail.com>

samsonasik commented on src/Definition.php:
```diff
@@ -127,57 +108,36 @@ public function hasMethod($method)
      *
      * @param  string $method
-     * @return null|\Laminas\Server\Method\Definition
+     * @return false|Method\Definition
```
`false` is not a type, use `bool` instead.

arueckauer replied:
Done.

arueckauer committed:
Fix return type in DocBlock

Signed-off-by: Andi Rückauer <arueckauer@gmail.com>

arueckauer committed:
Update documentation for 3.0
Add versioned structure
Move previous docs to v2
Add documentation structure for v3
Add v3 migration guide

Signed-off-by: Andi Rückauer <arueckauer@gmail.com>

froschdesign commented on mkdocs.yml:
```diff
@@ -1,11 +1,23 @@
+    base_url: https://docs.laminas.dev/
+    project_url: https://docs.laminas.dev/components/
+    current_version: v3
+    version:
```
```suggestion
    versions:
```

arueckauer replied:
Done.

froschdesign commented on mkdocs.yml:
```diff
@@ -1,11 +1,23 @@
 site_dir: docs/html
+extra:
+    project: Components
+    base_url: https://docs.laminas.dev/
+    project_url: https://docs.laminas.dev/components/
```
Can be removed because the build script sets these options.

arueckauer replied:
Done.

arueckauer committed:
Fix typo

Signed-off-by: Andi Rückauer <arueckauer@gmail.com>

arueckauer committed:
Remove base_url and project_url which get set by build script

Signed-off-by: Andi Rückauer <arueckauer@gmail.com>

michalbundyra commented on docs/book/v3/reference/migration.md:
```diff
@@ -0,0 +1,17 @@
+## Strict Typing
+
+Argument type declarations have been added. Passing a wrong argument type which
+previously might have worked, will not raise a TypeError.
```
typo: "will now"?

arueckauer replied:
Good catch.

michalbundyra commented on src/Definition.php:
```diff
@@ -102,23 +89,17 @@ public function addMethods(array $methods)
      *
-     * @param  array $methods Array of \Laminas\Server\Method\Definition objects or arrays
-     * @return \Laminas\Server\Definition
+     * @param  Method\Definition[]
```
missing param name - should be `$methods`

arueckauer replied:
Fixed.

michalbundyra commented on src/Definition.php:
```diff
@@ -6,59 +6,48 @@
-     * @param  null|array $methods
-     */
-    public function __construct($methods = null)
+    public function __construct(?array $methods = null)
```
note for later: we might want to change it to `iterable` type, instead of array

arueckauer replied:
Agreed.

michalbundyra commented on src/Method/Definition.php:
```diff
@@ -206,9 +167,9 @@ public function getMethodHelp()
      *
      * @param  object $object
```
as typehint has been added on the parameter this is redundant now

arueckauer replied:
Removed redundant @param and @return tags where applicable.

michalbundyra commented on src/Method/Prototype.php:
```diff
@@ -140,29 +111,22 @@ public function getParameterObjects()
      *
      * @param  string|int $index
-     * @return null|\Laminas\Server\Method\Parameter
+     * @return null|Parameter
```
no additional value in this return docblock, can be removed.

arueckauer replied:
Removed.

michalbundyra commented on src/Reflection.php:
```diff
@@ -59,27 +60,26 @@ public static function reflectClass($class, $argv = false, $namespace = '')
-    {
+    public static function reflectFunction(
+        $function,
+        ?array $argv = null,
```
Here we have also signature change - default value before was `false`, now is `null`.
We should note it in migration guide, as it is not just adding strict type.

Any signature changes should be noted in migration guide.

arueckauer replied:
Note added to the docs.

michalbundyra commented on src/Reflection.php:
```diff
@@ -59,27 +60,26 @@ public static function reflectClass($class, $argv = false, $namespace = '')
+    public static function reflectFunction(
+        $function,
+        ?array $argv = null,
+        ?string $namespace = null
```
Here is slightly different situation, but also signature change - default value from empty string to `null`.
We need to track down what should be the correct value, and how the code works with `null` and empty string.

arueckauer replied:
Note added to the docs.

This is a change that follows the signature change made in `AbstractFunction`, where $namespace is optional and thus nullable.

michalbundyra commented on src/Reflection/AbstractFunction.php:
```diff
@@ -91,13 +88,10 @@ abstract class AbstractFunction
      * @throws Exception\RuntimeException
      */
-    public function __construct(ReflectionFunctionAbstract $r, $namespace = null, $argv = [])
+    public function __construct(ReflectionFunctionAbstract $r, ?string $namespace = null, ?array $argv = null)
```
Signature change: default value `[]` -> `null`. Not sure why? I see default value in the property is `[]` so we can remove it, keep just array here, and remove the check `is_array` below.

arueckauer replied:
Totally makes sense. 👍🏻 Fixed.

michalbundyra commented on src/Reflection/AbstractFunction.php:
```diff
@@ -178,9 +166,9 @@ protected function buildTree()
      * @param array $paramTypes Array of arguments (each an array of types)
      * @param array $paramDesc Array of parameter descriptions
-     * @return array
+     * @return void
```
redundant, as we have RTH defined

arueckauer replied:
Removed.

michalbundyra commented on src/Reflection/AbstractFunction.php:
```diff
@@ -364,64 +349,46 @@ public function __get($key)
-     * @return void
      */
-    public function setNamespace($namespace)
+    public function setNamespace(?string $namespace = null): void
```
probably we should not have the default value here, we want to have `setNamespace` called always with the value.
Calling setter without any value is always confusing.

arueckauer replied:
Agreed. Remove default value. Additionally set `$namespace` to `null` if empty, since this would mean an non-present state.

michalbundyra commented on src/Reflection/AbstractFunction.php:
```diff
@@ -364,64 +349,46 @@ public function __get($key)
-    public function setNamespace($namespace)
+    public function setNamespace(?string $namespace = null): void
     {
         if (empty($namespace)) {
```
Because of that, maybe we should keep namespace as `string` and default (in properties) as empty string?
I think it was like that before. I am fine also with `null` which means that is not set/not defined, but it should behave the same as empty string probably...

arueckauer replied:
Yes, namespace can be optional. Not set is represented by `null`.

michalbundyra commented on src/Reflection/ReflectionClass.php:
```diff
@@ -122,47 +116,34 @@ public function __get($key)
-     * @return void
      */
-    public function setNamespace($namespace)
+    public function setNamespace(?string $namespace = null): void
```
the same as before, we should not have default value here, we should decide if we want to support null, or just empty string for unsetting the namespace

arueckauer replied:
Agreed. Default value removed.

michalbundyra commented on src/Reflection/AbstractFunction.php:
```diff
@@ -364,64 +349,46 @@ public function __get($key)
+    public function setNamespace(?string $namespace = null): void
     {
         if (empty($namespace)) {
             $this->namespace = '';
```
I see it in the other setter, that you changed the line to `null`, but not above.
Again: we should decide if we want to support null values, if so, probably here we should have `null` instead.

arueckauer replied:
Not sure, what you mean here. Can you elaborate?

arueckauer replied:
As discussed in chat: Fixed by setting `$namespace` to `null` if new value is empty.

michalbundyra commented on src/Reflection/ReflectionMethod.php:
```diff
@@ -26,20 +27,16 @@ class ReflectionMethod extends AbstractFunction
+        ReflectionClass $class,
+        \ReflectionMethod $r,
+        ?string $namespace = null,
+        ?array $argv = null
```
signature change, probably we want to keep just array, without nulls

arueckauer replied:
That is correct. Fixed.

michalbundyra commented on src/Reflection/ReflectionMethod.php:
```diff
@@ -93,13 +85,10 @@ public function __wakeup()
     {
         $docComment = $this->reflection->getDocComment();
-        if (strpos($docComment, self::INHERIT_TAG) !== false) {
+        if (false !== $docComment && strpos($docComment, self::INHERIT_TAG) !== false) {
```
is it some kind of bugfix here?

arueckauer replied:
Yes. `reflectionMethod()` possibly can return a non-string value (`false`), which causes a type error.

arueckauer replied:
The bug is only present with strict typing, which is why it never occured in `2.0`. IMO there is not bugfix needed for previous version, where strict typing is not supported with PHP ^5.0.

michalbundyra commented on src/Reflection/ReflectionMethod.php:
```diff
@@ -143,29 +129,19 @@ function ($docComment) {
         $parentReflectionClass = $reflectionClass->getParentClass();
         if (! $parentReflectionClass) {
-            return;
+            return null;
```
not to be changed in this PR, but do we want to keep null with arrays?
Usually it's easier to return empty array, so then we can iterate on the array etc.

arueckauer replied:
I like the suggested approach. Always returning arrays would make it easier.

michalbundyra commented on src/Reflection/ReflectionParameter.php:
```diff
@@ -88,89 +82,46 @@ public function __call($method, $args)
-     * @return void
-     */
-    public function setType($type)
+    public function setType(?string $type = null): void
```
remove default value from setter

also, i am a bit confused, as in constructor we do not allow nulls, so not sure if we should do it here?

arueckauer replied:
Removed, accepting only values of type `string`.

michalbundyra commented on src/Reflection/ReflectionParameter.php:
```diff
@@ -88,89 +82,46 @@ public function __call($method, $args)
-     * @return void
-     */
-    public function setDescription($description)
+    public function setDescription(?string $description = null): void
```
the same as on `setType` setter

arueckauer replied:
Since `$description` is optional, `null` is allowed and represents a non-present state of `$description`. I updated the constructor to also allow `null`.

michalbundyra commented on src/Reflection/ReflectionParameter.php:
```diff
@@ -88,89 +82,46 @@ public function __call($method, $args)
-     * @return void
-     */
-    public function setPosition($index)
+    public function setPosition(?int $index = null): void
```
also, the same here: do we want nullable types? please remove default param type on setter.

arueckauer replied:
Done, accepting only `int` values now.

michalbundyra commented on src/Reflection/ReflectionReturnValue.php:
```diff
@@ -16,78 +18,38 @@
-     * @return void
-     */
-    public function setType($type)
+    public function setType(?string $type = null): void
```
the same comments as in the previous class here

arueckauer replied:
Changed.

michalbundyra commented on src/Server.php:
```diff
@@ -20,12 +19,12 @@ interface Server
      * @return void
      */
-    public function addFunction($function, $namespace = '');
+    public function addFunction(string $function, ?string $namespace = null): void;
```
signature change - default value, here and the next couple more methods below

if it is something needed it should be documented in migration guide

arueckauer replied:
Note has been added to the migration guide.

michalbundyra reviewed (changes requested):
@arueckauer I've checked the PR and noted some issues.

There is more issues about redundant doc comment, but it's not super important and we can always polish it later. The most important here is that we really cannot change signature (default value false -> null, or [] -> null, etc)

See my comments and if you have any questions just let me know. 

Thanks for your effort! 👍

arueckauer committed:
Update DocBlocks
Use single-line comments for @var tags of class properties
Remove @param and @return tags which are made obsolete by type declarations
Remove class header comment which do not add valuable information
Add line breaks in DocBlocks in preparation for laminas-coding-standard:2.0
Use self as return type where applicable

Signed-off-by: Andi Rückauer <arueckauer@gmail.com>

froschdesign commented on docs/book/v3/reference/migration.md:
```diff
@@ -0,0 +1,17 @@
+following that you should be aware of, and potentially update your application
+to adopt:
+
+- [PHP 7.2 support](#php-7.2-support)
+- [Strict typing](#strict-typing)
```
Please remove these links because the documentation theme creates already a table of contents with links.

Example: https://docs.laminas.dev/laminas-config/intro/

arueckauer replied:
Removed.

arueckauer committed:
Add function imports for used functions
Add braces to new object instantiations

Signed-off-by: Andi Rückauer <arueckauer@gmail.com>

froschdesign commented on mkdocs.yml:
```diff
@@ -1,11 +1,21 @@
-    - Intro: intro.md
-    - Reflection: reflection.md
+    - v3:
+        - Intro: v3/intro.md
```
```suggestion
        - Introduction: v3/intro.md
```

arueckauer replied:
Done.

froschdesign commented on mkdocs.yml:
```diff
@@ -1,11 +1,21 @@
+        - Reference:
+              - Migration: v3/reference/migration.md
+    - v2:
+        - Intro: v2/intro.md
```
```suggestion
        - Introduction: v2/intro.md
```

arueckauer replied:
Changed.

arueckauer committed:
Fix typo to clarify TypeError will be thrown

Signed-off-by: Andi Rückauer <arueckauer@gmail.com>

arueckauer committed:
Add missing variable name

Signed-off-by: Andi Rückauer <arueckauer@gmail.com>

arueckauer committed:
Allow null values for  in constructor

Signed-off-by: Andi Rückauer <arueckauer@gmail.com>

arueckauer committed:
Remove default value on  setter

Signed-off-by: Andi Rückauer <arueckauer@gmail.com>

arueckauer committed:
Remove default value in  setter

Signed-off-by: Andi Rückauer <arueckauer@gmail.com>

arueckauer committed:
Remove default value in  setter

Signed-off-by: Andi Rückauer <arueckauer@gmail.com>

arueckauer committed:
Remove default value in  setter

Signed-off-by: Andi Rückauer <arueckauer@gmail.com>

arueckauer committed:
Add note about signature change in Server interface

Signed-off-by: Andi Rückauer <arueckauer@gmail.com>

arueckauer committed:
Remove toc links, since they are generated automatically

Signed-off-by: Andi Rückauer <arueckauer@gmail.com>

arueckauer committed:
Change page title to Indtroduction

Signed-off-by: Andi Rückauer <arueckauer@gmail.com>

arueckauer committed:
Change page title to Indtroduction

Signed-off-by: Andi Rückauer <arueckauer@gmail.com>

arueckauer committed:
Add note about signature change of  in Reflection::reflectionClass

Signed-off-by: Andi Rückauer <arueckauer@gmail.com>

arueckauer committed:
Add Reflection::reflectionFunction mention to signature change

Signed-off-by: Andi Rückauer <arueckauer@gmail.com>

arueckauer committed:
Add note about signature change of  argument type for 
eflectClass and 
eflectFunction in Reflection

Signed-off-by: Andi Rückauer <arueckauer@gmail.com>

arueckauer committed:
Revert default value change, making is_array() check obsolete

Signed-off-by: Andi Rückauer <arueckauer@gmail.com>

arueckauer committed:
Remove default value for  in setter
If empty, namespace is not present, thus null

Signed-off-by: Andi Rückauer <arueckauer@gmail.com>

arueckauer committed:
Remove default value for  in setter
Update constructor to changed setter behaviour

Signed-off-by: Andi Rückauer <arueckauer@gmail.com>

arueckauer committed:
Revert type change of  in ReflectionMethod
Only allow values of type array

Signed-off-by: Andi Rückauer <arueckauer@gmail.com>

arueckauer committed:
Add assertion to verify save returns false on a non-existent file

Signed-off-by: Andi Rückauer <arueckauer@gmail.com>

arueckauer committed:
Add assertion to verify get and delte return false on a non-existent file

Signed-off-by: Andi Rückauer <arueckauer@gmail.com>

arueckauer committed:
Add assertions to ensure empty string argument for  setter, sets  to null

Signed-off-by: Andi Rückauer <arueckauer@gmail.com>

arueckauer committed:
Add assertions to verify InvalidArgumentException is thrown for invalid namespace values

Signed-off-by: Andi Rückauer <arueckauer@gmail.com>

samsonasik commented on docs/book/v3/intro.md:
```diff
@@ -0,0 +1,13 @@
+# Introduction
+
+The laminas-server family of classes provides functionality for the various server
```
I think it needs "is a" and need to change the "family" to "collection"

```suggestion
The laminas-server is a collection of classes provides functionality for the various server
```

arueckauer replied:
The text is based on README, where the same expression is being used. Personally, I don't see the necessity of a change. Collection sounds much larger than just 2 (JSON-RPC and XML-RPC) packages.

samsonasik replied:
ok, I think the "is a" is still needed.

arueckauer replied:
I am not following. "provides" is the verb in the sentence.

samsonasik replied:
ok 👍

samsonasik commented on src/AbstractServer.php:
```diff
@@ -6,45 +6,31 @@
 use ReflectionClass;
+use ReflectionException;
+
+use function call_user_func_array;
```
can use invoked function instead.

samsonasik replied:
oh, just see your comment https://github.com/laminas/laminas-server/pull/6/files#r421267551

samsonasik commented on src/Definition.php:
```diff
@@ -6,94 +6,86 @@
         foreach ($methods as $key => $method) {
+            if (is_numeric($key)) {
+                $key = null;
+            }
```
`is_numeric` is no need here, as next line is calling `addMethod()` which has `is_numeric` check agains key as a name, make it unneded double check.

arueckauer replied:
It is required, because a numeric (integer) value passed to `addMethod()` will throw a TypeError.

Array keys of a numeric string type (e.g. `'100'`) will be cast to an integer internally by PHP. This case is covered by `DefinitionTest::testConstructorNumericKeyWillBeReplacedByMethodName()`.

samsonasik commented on src/Method/Definition.php:
```diff
@@ -172,43 +138,30 @@ public function setPrototypes(array $prototypes)
-    public function setObject($object)
+    public function setObject(object $object): self
     {
         if (! is_object($object) && (null !== $object)) {
```
`is_object()` check is not needed here since the parameter required is object

arueckauer replied:
Agreed and fixed.

samsonasik commented on src/Method/Definition.php:
```diff
@@ -172,43 +138,30 @@ public function setPrototypes(array $prototypes)
-     * @return \Laminas\Server\Method\Definition
      */
-    public function setObject($object)
+    public function setObject(object $object): self
```
`null` need to be allowed since the `getObject()` allow null returns value.

arueckauer replied:
Getter return type must be nullable, since property might be uninitialized. "Resetting" an object to null makes only sense, if a previously set object needs to be reverted. Is there a use case for that?

samsonasik replied:
ok 👍 , make sense

arueckauer committed:
Remove type validation which is made obsolete by argument type declaration

Signed-off-by: Andi Rückauer <arueckauer@gmail.com>

samsonasik commented on src/Method/Definition.php:
```diff
@@ -172,92 +138,47 @@ public function setPrototypes(array $prototypes)
-     * @return null|object
-     */
-    public function getObject()
+    public function getObject(): ?object
```
if null is allowed to be returned, the `setObject()` need to allow null parameter.

arueckauer replied:
There is already a discussion about that on the setter...

arueckauer committed:
Remove no longer used imports

Signed-off-by: Andi Rückauer <arueckauer@gmail.com>

weierophinney commented on docs/book/v3/intro.md:
```diff
@@ -0,0 +1,13 @@
+
+The laminas-server family of classes provides functionality for the various server
+classes, including [laminas-xmlrpc](https://docs.laminas.dev/laminas-xmlrpc) and
+[laminas-json-server](https://docs.laminas.dev/laminas-json-server/).
```
And laminas-soap!

arueckauer replied:
Added `laminas-soap` to v3 Introduction and README.

weierophinney commented on docs/book/v3/reference/migration.md:
```diff
@@ -0,0 +1,34 @@
+# Migration to version 3.0
+
+However, starting in this version, we offer a few changes affecting the
```
```suggestion
Starting in version 3.0, we offer a few changes affecting the
```

arueckauer replied:
✔

weierophinney commented on docs/book/v3/reference/migration.md:
```diff
@@ -0,0 +1,34 @@
+## Strict Typing
+
+Argument type declarations have been added. Passing a wrong argument type which
+previously might have worked, will now raise a TypeError.
```
```suggestion
previously might have worked will now raise a `TypeError`.
```

arueckauer replied:
✔

weierophinney commented on docs/book/v3/reflection.md:
```diff
@@ -0,0 +1,51 @@
+
+`reflectFunction()` returns a `Laminas\Server\Reflection\Function` object;
+`reflectClass()` returns a `Laminas\Server\Reflection\Class` object. Please refer
+to the API documentation to see what methods are available to each.
```
The one problem with this statement is that we do not publish the API documentation. Maybe point them to the source code?

arueckauer replied:
This is a copy of the v2 document. Since logical behavior has not changed, I did not update this document yet. I agree, it does not make sense referencing something that does not exist. I added the links to the source.

weierophinney reviewed (approved):
This looks great!

I've provided some suggestions for rewording a few items in the documentation, but otherwise, I think this is ready. I'm looking forward to the next batch of changes, to see what additional things we can simplify and improve!

arueckauer committed:
Add laminas-soap to list of laminas-server family
Order members alphabetically

Signed-off-by: Andi Rückauer <arueckauer@gmail.com>

arueckauer committed:
Improve reading experience and understanding in documentation

Signed-off-by: Andi Rückauer <arueckauer@gmail.com>

arueckauer committed:
Add improvements to documentation for a better reading experience and understanding

Signed-off-by: Andi Rückauer <arueckauer@gmail.com>

michalbundyra commented:
@weierophinney What about changelog entry for these changes?

weierophinney commented:
> What about changelog entry for these changes?

Ugh, forgot - even though I remembered on the other PRs that @arueckauer submitted today.

michalbundyra commented:
@weierophinney 

> Ugh, forgot ...

Still, we need to add it then :)

weierophinney commented:
> Still, we need to add it then :)

Done in 9aef525
